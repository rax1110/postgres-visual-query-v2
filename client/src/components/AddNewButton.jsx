import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'react';
import * as PropTypes from 'prop-types';
import { Button } from 'reactstrap';

export const AddNewButton = ({ id, size, onClick }) => (
  <Button
    className=""
    outline
    color="info"
    size={size}
    id={id}
    onClick={onClick}
  >
    <FontAwesomeIcon icon="plus" />
  </Button>
);

AddNewButton.propTypes = {
  id: PropTypes.string,
  size: PropTypes.string,
  onClick: PropTypes.func,
};
