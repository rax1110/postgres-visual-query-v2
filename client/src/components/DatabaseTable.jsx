import React from 'react';
import { Button, Tooltip } from 'reactstrap';
import { connect } from 'react-redux';
import _ from 'lodash';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { addColumn, addTable, removeTable } from '../actions/queryActions';
import { withToggle } from '../hocs/withToggle';

export const DatabaseTable = ({ data, checked, id, addTableProp, toggle, toggleStatus }) => {
  const handleOnClick = () => {
    addTableProp(data);
  };

  let tableTypeColor = 'primary';

  if (_.isEqual(data.table_type, 'VIEW')) {
    tableTypeColor = 'info';
  } else if (_.isEqual(data.table_type, 'FOREIGN')) {
    tableTypeColor = 'secondary';
  } else if (_.isEqual(data.table_type, 'MATERIALIZED VIEW')) {
    tableTypeColor = 'light';
  }

  const btnSelected = checked ? 'success' : tableTypeColor;

  const modifiers = {
    preventOverflow: {
      enabled: false,
    },
    hide: {
      enabled: false,
    },
  };

  return (
    <div className="w-100 pr-1">
      <Button
        size="sm"
        color={btnSelected}
        id={id}
        className="btn-block my-1 pt-0 text-left"
        onClick={handleOnClick}
      >
        <small
          color={tableTypeColor}
          className="text-truncate align-self-start"
        >
          <span className="mr-1 px-0">
            <FontAwesomeIcon icon="th-large" />
          </span>
          {data.table_type}
        </small>
        <div className="text-truncate">
          {` ${data.table_name}`}
        </div>
      </Button>
      <Tooltip
        placement="right"
        isOpen={toggleStatus}
        target={id}
        toggle={toggle}
        modifiers={modifiers}
        delay={{ show: 200, hide: 0 }}
        className=""
      >
        {data.table_name}
      </Tooltip>
    </div>
  );
};

DatabaseTable.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  checked: PropTypes.bool,
  id: PropTypes.string,
  addTableProp: PropTypes.func,
  toggle: PropTypes.func,
  toggleStatus: PropTypes.bool,
};

const mapDispatchToProps = {
  addColumn,
  addTableProp: data => addTable(data),
  removeTable,
};

export default withToggle(connect(null, mapDispatchToProps)(DatabaseTable));
