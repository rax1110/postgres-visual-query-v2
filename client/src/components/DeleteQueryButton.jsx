import React from 'react';
import { Button } from 'reactstrap';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { deleteQuery, setActiveQuery } from '../actions/queryActions';
import { deleteQueries, removeMainFromQueries } from '../actions/queriesActions';

export const DeleteQueryButton = (props) => {
  const handleOnClick = () => {
    props.deleteQuery();

    const mainQuery = props.queries.find(query => query.id === 0);

    if (mainQuery) {
      props.removeMainFromQueries();
      props.setActiveQuery(mainQuery);
    } else {
      props.deleteQueries();
    }
  };

  return (
    <Button color="danger" className="mr-2" onClick={handleOnClick}>
      <FontAwesomeIcon icon="trash" />
    </Button>
  );
};

DeleteQueryButton.propTypes = {
  deleteQuery: PropTypes.func,
  deleteQueries: PropTypes.func,
  queries: PropTypes.arrayOf(PropTypes.shape({})),
  removeMainFromQueries: PropTypes.func,
  setActiveQuery: PropTypes.func,
};

const mapStateToProps = store => ({
  queries: store.queries,
});

const mapDispatchToProps = {
  deleteQuery,
  deleteQueries,
  removeMainFromQueries,
  setActiveQuery,
};

export default connect(mapStateToProps, mapDispatchToProps)(DeleteQueryButton);
