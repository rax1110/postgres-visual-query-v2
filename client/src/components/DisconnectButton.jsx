import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { Redirect } from 'react-router-dom';
import * as PropTypes from 'prop-types';
import { disconnect } from '../actions/hostActions';
import { translations } from '../utils/translations';

export const DisconnectButton = (props) => {
  const handleOnClick = () => {
    props.disconnect();
  };

  if (!props.connected) {
    return <Redirect to="/" />;
  }

  return (
    <Button size="lg" className="btn-block my-2 px-2" color="danger" onClick={handleOnClick}>
      {translations[props.language.code].sideBar.disconnectB}
    </Button>
  );
};

DisconnectButton.propTypes = {
  disconnect: PropTypes.func,
  connected: PropTypes.bool,
  language: PropTypes.shape({ code: PropTypes.string }),
};

const mapStateToProps = store => ({
  connected: store.host.connected,
  language: store.settings.language,
});

const mapDispatchToProps = {
  disconnect,
};

export default connect(mapStateToProps, mapDispatchToProps)(DisconnectButton);
