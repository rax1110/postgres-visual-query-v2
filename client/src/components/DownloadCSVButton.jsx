import React from 'react';
import { connect } from 'react-redux';
import { CSVLink } from 'react-csv';
import _ from 'lodash';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadCSVButton = (props) => {
  const getHeaders = (result) => {
    if (_.isNull(result)) {
      return [];
    }

    const headers = [];

    result.fields.forEach((field) => {
      const header = {
        label: field.name,
        key: field.name,
      };

      headers.push(header);
    });

    return headers;
  };

  const getData = (result) => {
    if (_.isNull(result)) {
      return [];
    }
    return result.rows;
  };

  const disabled = _.isNull(props.result);

  return (
    <CSVLink
      className="mr-2"
      data={getData(props.result)}
      headers={getHeaders(props.result)}
      filename="result.csv"
    >
      <Button disabled={disabled}>
        <FontAwesomeIcon icon="download" />
        <div className="d-inline"> CSV</div>
      </Button>
    </CSVLink>
  );
};

DownloadCSVButton.propTypes = {
  result: PropTypes.shape({}),
};

const mapStateToProps = store => ({
  result: store.query.result,
});

export default connect(mapStateToProps)(DownloadCSVButton);
