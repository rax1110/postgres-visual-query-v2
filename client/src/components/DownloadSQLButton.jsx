import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadSQLButton = (props) => {
  const downloadContent = (name, content) => {
    if (navigator.msSaveBlob) {
      const blobObject = new Blob([content], { type: 'text/plain' });

      window.navigator.msSaveOrOpenBlob(blobObject, name);
    } else {
      const aTag = document.createElement('a');
      const file = new Blob([content], { type: 'text/plain' });

      aTag.href = URL.createObjectURL(file);
      aTag.download = name;

      document.body.appendChild(aTag);
      aTag.click();
    }
  };

  return (
    <Button
      className="mr-2"
      onClick={() => downloadContent('select_query.sql', props.sql)}
    >
      <FontAwesomeIcon icon="download" />
      <div className="d-inline"> SQL</div>
    </Button>
  );
};

DownloadSQLButton.propTypes = {
  sql: PropTypes.string,
};

const mapStateToProps = store => ({
  sql: store.query.sql,
});

export default connect(mapStateToProps)(DownloadSQLButton);
