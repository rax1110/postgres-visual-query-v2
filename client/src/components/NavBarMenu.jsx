import * as PropTypes from 'prop-types';
import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Row } from 'reactstrap';
import _ from 'lodash';
import { InputWithDeleteButton } from './InputWithDeleteButton';
import { AddNewButton } from './AddNewButton';
import { translations } from '../utils/translations';
import { addQuery } from '../actions/queriesActions';
import { setActiveQuery } from '../actions/queryActions';

export const NavBarMenu = (props) => {
  const [queryName, setQueryName] = useState('');

  const handleAddQuery = () => {
    props.addQuery(props.activeQuery.id);
  };

  const handleChange = (e) => {
    setQueryName(e.target.value);

    let query = _.cloneDeep(props.activeQuery);

    query = {
      ...query,
      [e.currentTarget.name]: e.target.value || '',
    };

    props.setActiveQuery(query);
  };

  const handleRemove = (e) => {
    let query = _.cloneDeep(props.activeQuery);

    setQueryName('');

    query = {
      ...query,
      [e.currentTarget.name]: '',
    };

    props.setActiveQuery(query);
  };

  return (
    <div className="pl-2 align-self-start m-0 pt-1 pb-2 bg-light">
      {props.activeQuery.id !== 0 && (
        <>
          <Row form>
            <div className="col-auto">
              <InputWithDeleteButton
                className="pb-2"
                id="newQueryAlias"
                name="queryName"
                placeholder={`${translations[props.language.code].queryBuilder.queryNamePh}`}
                tooltipTarget="newQueryAlias"
                tooltipText={` ${translations[props.language.code].tooltips.queryName}`}
                value={queryName}
                handleChange={handleChange}
                handleRemove={handleRemove}
              />
            </div>
          </Row>
        </>
      )}
      <div className="col-12 pl-0 text-info pt-6">
        <AddNewButton size="sm" id="newQuery" onClick={handleAddQuery} />
        {` ${translations[props.language.code].queryBuilder.queryH}`}
      </div>
    </div>
  );
};

NavBarMenu.propTypes = {
  language: PropTypes.shape({ code: PropTypes.string }),
  activeQuery: PropTypes.shape({ id: PropTypes.number }),
  addQuery: PropTypes.func,
  setActiveQuery: PropTypes.func,
};

const mapStateToProps = store => ({
  activeQuery: store.query,
  queries: store.queries,
});

const mapDispatchToProps = {
  addQuery,
  setActiveQuery,
};

export default connect(mapStateToProps, mapDispatchToProps)(NavBarMenu);
