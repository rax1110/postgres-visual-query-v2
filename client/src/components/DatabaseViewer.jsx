import React from 'react';
import { connect } from 'react-redux';
import { Scrollbars } from 'react-custom-scrollbars';
import _ from 'lodash';
import * as PropTypes from 'prop-types';
import DatabaseTable from './DatabaseTable';
import { filterTable } from '../utils/filterTable';

export const DatabaseViewer = (props) => {
  const constructData = (table) => {
    const data = {
      table_schema: table.table_schema,
      table_name: table.table_name,
      table_type: table.table_type,
      table_alias: '',
    };

    let constraints = JSON.parse(JSON.stringify(props.constraints));

    constraints = constraints.filter(constraint => constraint.table_schema === data.table_schema
      && constraint.table_name === data.table_name);

    let columns = JSON.parse(JSON.stringify(props.columns));

    columns = columns.filter(column => column.table_name === data.table_name
      && column.table_schema === data.table_schema).map((column) => {
      const col = column;

      col.constraints = constraints.filter(
        constraint => _.includes(constraint.column_name, column.column_name),
      );

      delete col.table_name;
      delete col.table_schema;
      return col;
    });

    data.columns = columns;

    return data;
  };

  return (
    <div className="flex-fill">
      <Scrollbars className="d-flex" autoHide>
        <div className="mt-1 pr-2">
          {props.tables.map((table, index) => {
            const checked = props.queryTable.some(
              queryTable => _.isEqual(table.table_name, queryTable.table_name)
                && _.isEqual(table.table_schema, queryTable.table_schema),
            );

            const id = `database-table-${index}`;

            return table.table_schema === props.selectedSchema
              && filterTable(table, props.searchExpr)
              && (
                <DatabaseTable
                  data={constructData(table)}
                  checked={checked}
                  key={id}
                  id={id}
                />
              );
          })}
        </div>
      </Scrollbars>
    </div>
  );
};

DatabaseViewer.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  constraints: PropTypes.arrayOf(PropTypes.shape({})),
  columns: PropTypes.arrayOf(PropTypes.shape({})),
  tables: PropTypes.arrayOf(PropTypes.shape({})),
  queryTable: PropTypes.arrayOf(PropTypes.shape({})),
  selectedSchema: PropTypes.string,
  searchExpr: PropTypes.string,
};

const mapStateToProps = store => ({
  tables: store.database.tables,
  schemas: store.database.schemas,
  selectedSchema: store.database.selectedSchema,
  constraints: store.database.constraints,
  columns: store.database.columns,
  queryTable: store.query.tables,
  searchExpr: store.database.searchExpr,
});

export default connect(mapStateToProps)(DatabaseViewer);
