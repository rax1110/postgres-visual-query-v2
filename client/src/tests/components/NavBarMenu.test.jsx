import React from 'react';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import { NavBarMenu } from '../../components/NavBarMenu';

describe('Component: NavBarMenu', () => {
  let store;
  let props;
  let mockStore;
  let component;
  let div;

  beforeEach(() => {
    props = {
      language: { code: 'eng' },
      query: {
        id: 1,
        queryName: 'Query 1',
      },
      activeQuery: {
        id: 2,
        queryName: 'Main',
      },
      addQuery: jest.fn(),
      setActiveQuery: jest.fn(),
    };

    div = document.createElement('div');

    document.body.appendChild(div);

    mockStore = configureStore([thunk]);
    store = mockStore(props);
    component = mount(
      <Provider store={store}>
        <NavBarMenu {...props} />
      </Provider>, { attachTo: div },
    );
  });

  it('renders correctly', () => {
    store = mockStore(props);

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('NavBarMenu handleAddQuery calls addQuery once', () => {
    component.find('AddNewButton').prop('onClick')();

    expect(props.addQuery).toBeCalled();
  });

  test('NavBarMenu handleChange should set correct query name and call setActiveQuery once', () => {
    component.find('input').simulate('change', {
      target: { value: 'queryTest' },
    });

    expect(component.find('input').prop('value')).toEqual('queryTest');
    expect(props.setActiveQuery).toBeCalled();
  });

  test('NavBarMenu handleRemove should set empty query name and call setActiveQuery once', () => {
    component.find('button').first().simulate('click', {
      currentTarget: { name: 'queryTest' },
    });

    expect(component.find('input').first().prop('value')).toEqual('');
    expect(props.setActiveQuery).toBeCalled();
  });
});
