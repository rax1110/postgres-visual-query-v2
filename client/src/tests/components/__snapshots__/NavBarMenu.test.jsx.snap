// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: NavBarMenu renders correctly 1`] = `
<div
  className="pl-2 align-self-start m-0 pt-1 pb-2 bg-light"
>
  <div
    className="form-row"
  >
    <div
      className="col-auto"
    >
      <div
        className="pb-2 input-group input-group-sm"
      >
        <input
          className="text-dark w-auto form-control"
          id="newQueryAlias"
          name="queryName"
          onChange={[Function]}
          placeholder="Query name"
          style={
            Object {
              "flexBasis": "auto",
            }
          }
          type="text"
          value=""
        />
        <div
          className="input-group-append"
        >
          <button
            aria-label={null}
            className="btn btn-danger"
            id="newQueryAlias"
            name="queryName"
            onClick={[Function]}
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="col-12 pl-0 text-info pt-6"
  >
    <button
      aria-label={null}
      className="btn btn-outline-info btn-sm"
      id="newQuery"
      onClick={[Function]}
      type="button"
    />
     Add query
  </div>
</div>
`;
